lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1600,16 +1584,15 @@ static void checkGenericParamAccessibility(TypeChecker &TC,
     }
   }
 
-  if (minAccessScope) {
-    auto minAccess = accessibilityFromScopeForDiagnostics(minAccessScope);
-
+  if (!minAccessScope.isPublic()) {
+    auto minAccess = minAccessScope.accessibilityForDiagnostics();
     bool isExplicit =
       owner->getAttrs().hasAttribute<AccessibilityAttr>() ||
       owner->getDeclContext()->getAsProtocolOrProtocolExtensionContext();
     auto diag = TC.diagnose(owner, diag::generic_param_access,
                             owner->getDescriptiveKind(), isExplicit,
-                            contextAccess, minAccess,
-                            accessibilityErrorKind);
+                            contextAccess, minAccess, accessibilityErrorKind,
+                            minAccessScope.isFileScope());
```

Comment:
`isFileScope` is *always* true for `fileprivate`, so I don't think this is a sufficient test. We also need to know if the *owner* is at file scope (and similarly for all the other diagnostics). The ultimate goal here is to take the new warning path if marking the offending decl as `private` will actually work.

(I'm sad that the diagnostics are getting more complicated, but I think your approach is the best one.)